### ISSUE TITLE
S2187: Rule should not raise False Positives if MSTEST test method is in a base class higher up in the inheritance hierarchy

### ISSUE THREAD

**Issue by ThomasMader** [SonarSource/sonar-dotnet#5507]
### Description

I am getting S2187 if a test method is provided in an abstract class higher up the inheritance hierarchy only.

### Repro steps

```
	[TestClass]
	public abstract class A
	{
		[TestMethod]
		public void Test()
		{
		}
	}

	[TestClass]
	public abstract class B : A
	{
		// works if this is uncommented
		//[TestMethod]
		//public void Test()
		//{
		//}
	}

	[TestClass]
	public class C : B
	{

	}
```

### Expected behavior

No warning.

### Actual behavior

Warning about S2187 on class C.

### Known workarounds

Suppress warning

### Related information

SonarLint 5.5.0.43817
SonarQube 8.9.6.50800
Visual Studio 17.1.2

**Comment by costin-zaharia-sonarsource** [comment]
Hi @ThomasMader, 

I'm not sure I understand what is the use case for this. In your example, both `C` and `B` classes can be removed and `A` can be made non-abstract.

**Comment by ThomasMader** [comment]
@costin-zaharia-sonarsource yes that is correct the code I provided clearly doesn't make much sense it's just the most minimal code I could write to reproduce the issue.

In our code base we have two abstract classes in this case because there are two different base classes we want to provide.
